packages/typegpu/src/core/variable/tgpuVariable.ts

```diff
@@ -0,0 +1,98 @@
+import type { AnyWgslData } from '../../data/wgslTypes';
+import { inGPUMode } from '../../gpuMode';
+import type { TgpuNamable } from '../../namable';
+import type { Infer } from '../../shared/repr';
+import type { ResolutionCtx, TgpuResolvable } from '../../types';
+
+// ----------
+// Public API
+// ----------
+
+export type VariableScope = 'private' | 'workgroup';
+
+export interface TgpuVar<TDataType extends AnyWgslData>
+  extends TgpuResolvable,
+    TgpuNamable {
+  value: Infer<TDataType>;
+  readonly scope: VariableScope;
+}
+
+export interface TgpuVarUninitialized<TDataType extends AnyWgslData>
+  extends TgpuVar<TDataType> {
+  $scope(scope: VariableScope): this;
+}
```

Comment:
Do we need to have two different interfaces? If we want to represent initialization on the type level I would opt for something more like `TgpuVar<...> & Initialized`. I also think I would rather have the scope be represented in the type rather than initialization status (as a user).